src/src/PluginStructs/P145_data_struct.cpp

```diff
@@ -0,0 +1,598 @@
+      0.0,          // CORD
+      0.0,          // CORE
+      0.0,          // CORF
+      0.0,          // CORG
+      p145AlgB,     // preferred/tuned algorithm
+      "MQ-4",       // Name
+      "CH4",        // gas
+  },
+  // MQ-7
+  {
+      27.5,         // cleanRatio
+      491204,       // PARA scaling
+      -5.826,       // PARB exponent
+      0.0,          // CORA
+      0.0,          // CORB
+      0.0,          // CORC
+      0.0,          // CORD
+      0.0,          // CORE
+      0.0,          // CORF
+      0.0,          // CORG   
+      p145AlgB,     // preferred/tuned algorithm
+      "MQ-7",       // Name
+      "CO",         // gas
+  }
+};
+/// @brief The number of types stored in the sensorData[] table
+constexpr const int nbrOfTypes = (int)(sizeof(sensorData) / sizeof(struct P145_SENSORDEF));
+
+/*****************************************************************************/
+/*!
+@brief  Perform the calibration algorithm
+@param[in] currentRcal  Value for Rzero determined with current measurement
+@return Void
+@note   Collect the highest Rzero during the calibration period
+        At the end of the period take mean value of current and calculated Rzero
+*/
+/*****************************************************************************/
+void P145_data_struct::calibrate (float currentRcal)
+{
+  unsigned long now = millis();
+  long time = timePassedSince(last_cal);
+  float lastRcal = rcal;  // Last calculated Rcal this calibration sequence
+  
+  if (currentRcal > rcal)
+  {
+    rcal = currentRcal;
+  }
+  const bool doit = (time > P145_CALIBRATION_INTERVAL) && (rcal > 0.0f);
+  if (doit)
+  {
+    rzero = rcal;     // Update Rzero as determined by calibration
+    rcal = 0.0;       // Restart calibration cycle with no value
+    last_cal = now;   // Remember calibration moment
+  }
+
+  if (loglevelActiveFor(LOG_LEVEL_INFO))
+  {
+#ifdef P145_DEBUG
+    // Calculated Rzero if calibration concentration is applied
+    // Rcal as calculated previous sample
+    addLog(LOG_LEVEL_INFO, concat(concat(F("MQ-xx: Calibration with Rcal =  "), currentRcal), concat(F(" Rlast = "), lastRcal))); 
+#endif
+    if (doit)
+    {
+      addLog(LOG_LEVEL_INFO, concat(F("MQ-xx: ***Calibrating*** Rzero =  "), rcal));
+    }
+  }
+}
+
+/*****************************************************************************/
+/*!
+@brief  Get the resistance of the sensor, ie. the property to be measured
+@param[in] val     Analog input value read from ADC [counts]
+@return The sensor resistance in Ohm
+@note   Uses hard coded constants:
+        MAX_ADC_VALUE Max range for ADC
+        VMAX          Analog  input voltage corresponding to MAX_ADC_VALUE
+        VCC           Voltage applied to the sensor-Rload combination
+*/
+/*****************************************************************************/
+float P145_data_struct::getResistance(float val) const
+{
+  return ((MAX_ADC_VALUE * P145_VCC) / (P145_VMAX * val) - 1.0f) * rload;
```

Comment:
What if `val` equals 0?

Reply:
What I have seen is that the formula is executed with a result of "inf". No crashes, but value out of range. It is debatable if we should rely on this or proactive check for boundaries. Problem is what happens if it is very close to zero...
I will add boundary checking at the cost of some extra code.

Reply:
See the functions in `src/Helpers/ESPEasy_math.h` :)